Remove null from model. Replace with "field"

awong-dev opened the pull request:
Sets the stage for doing validations correctly.

akainic commented on _health-care/_js/components/form-elements/ErrorableTextInput.jsx:
```diff
@@ -56,7 +62,7 @@ class ErrorableTextInput extends React.Component {
             placeholder={this.props.placeholder}
             type="text"
-            value={this.props.value}
+            value={this.props.value.value}
```
I wonder if we should change the prop name `value` to be something else? Maybe `field` or `fieldData`? Since what used to be `value` is now a function that returns 2 different things, an actual `value` and `dirty`.

awong-dev replied:
Yes...I was going to go for `field` in the next pass.

akainic commented on _health-care/_js/reducers/veteran/index.js:
```diff
@@ -2,7 +2,7 @@ import _ from 'lodash';
 import lodashDeep from 'lodash-deep';
 
 import { ENSURE_FIELDS_INITIALIZED, VETERAN_FIELD_UPDATE } from '../../actions';
-import { initializeNullValues } from '../../utils/validations';
```
We can probably delete this function now from `validations.js`.

akainic commented on _health-care/_js/reducers/veteran/index.js:
```diff
@@ -13,171 +13,171 @@ _.mixin(lodashDeep);
-    maritalStatus: null,
-    sectionComplete: false
+    maritalStatus: makeField(''),
+    sectionComplete: makeField(false)
```
We might not want to change `sectionComplete` from just being `false`, since this is currently breaking form styles.

awong-dev committed:
Remove null from model. Replace with "field"

The field object has a value and a dirty bit that indicates if the field
has never been touched before. This allows us to disable validation on
the initial render.

Unforutnately, fixing this also meant changing nearly every file.

akainic commented on _health-care/_js/utils/validations.js:
```diff
@@ -107,14 +101,50 @@ function isValidAddress(street, city, country, state, zipcode) {
+}
+
+function isValidRequiredField(validator, field) {
+  return isNotBlank(field.value) || validator(field.value);
```
I think this should be `&&` instead of `||`?

awong-dev replied:
fixed and unittested.

akainic commented on _health-care/_js/utils/validations.js:
```diff
@@ -107,14 +101,50 @@ function isValidAddress(street, city, country, state, zipcode) {
+  return isValidFullNameField(data.fullName) &&
+      isValidRequiredField(isValidSSN, data.socialSecurityNumber) &&
+      isNotBlank(data.gender.value) &&
       isNotBlank(data.maritalStatus) &&
```
Should this be passed `data.maritalStatus.value` instead?

awong-dev replied:
yup. done.

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableSelect.spec.jsx:
```diff
@@ -19,53 +21,53 @@ describe('<ErrorableSelect>', () => {
     });
 
-    it('label must be a string', () => {
+    it('label must be a stringthis.props.', () => {
```
Mistake?

awong-dev replied:
fixed

akainic commented on spec/javascripts/health-care/components/form-elements/ErrorableTextInput.spec.jsx:
```diff
@@ -17,47 +18,47 @@ describe('<ErrorableTextInput>', () => {
     it('label is required', () => {
       SkinDeep.shallowRender(
-        <ErrorableTextInput onValueChange={(_update) => {}}/>);
+        <ErrorableTextInput field={makeField(1)} onValueChange={(_update) => {}}/>);
```
Since we have a different component for number inputs, perhaps we should pass this a string instead of a number?

awong-dev replied:
yah...you're right. fixed.

akainic commented on _health-care/_js/components/form-elements/DateInput.jsx:
```diff
@@ -29,33 +29,33 @@ class DateInput extends React.Component {
     } else {
-      isValid = (isBlank(this.props.day) && isBlank(this.props.month) && isBlank(this.props.year)) ||
-          isValidDate(this.props.day, this.props.month, this.props.year);
+      isValid = (isBlank(day) && isBlank(month) && isBlank(year)) ||
```
can we use `isBlankDateField` instead?

akainic commented on _health-care/_js/components/form-elements/ErrorableSelect.jsx:
```diff
@@ -93,7 +95,10 @@ ErrorableSelect.propTypes = {
     ])).isRequired,
   required: React.PropTypes.bool,
-  value: React.PropTypes.string,
+  value: React.PropTypes.shape({
```
Why don't we want to change this prop to being called `field` instead as well?

awong-dev replied:
Would like to save for another pass. This PR is too big already.

akainic commented on _health-care/_js/components/personal-information/VeteranAddressSection.jsx:
```diff
@@ -46,16 +46,16 @@ class VeteranAddressSection extends React.Component {
-              value={this.props.data.county}
-              onValueChange={(update) => {this.onStateChange('county', update);}}/>
+              field={this.props.data.county}
+              onValueChange={(update) => {this.props.onStateChange('county', update);}}/>
```
Oof, good catch

akainic commented:
@awong-dev Reviewed and back to you!

awong-dev committed:
Respond to Anne's comments.

akainic commented:
Something a little strange is happening when I test this locally. In `NameAndGeneralInfo`, `gender`, and `maritalStatus` are loading with error styling. Debugging locally now to try to figure out what that is happening.

akainic commented:
I think I figured out why the above is happening: something was added to `FullName.jsx` and `SocialSecurityNumber.jsx` to check if the field is dirty in the validation. If not, it returns true, which is why there are no errors on these fields on initial load. For `Gender.jsx`, we are passing the wrong thing to the validation function, which is why it is not displaying errors on load or on continue. We are passing `this.props.value`, but it should be `this.props.value.value`. `DateInput` and `maritalstatus` are not checking if the field is dirty before running validations, which is why they are showing errors on load.

akainic commented:
Also weird things are happening with the validations throughout the form. Sorry, this is going to be really tedious to fix :/. I'm happy to take on that work.

awong-dev commented:
migrated to https://github.com/department-of-veterans-affairs/healthcare-application/pull/10
